textcursor.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gxui
+
+import "github.com/google/gxui/interval"
+
+// TextCursor represents a text position and selection. A selection is defined as
+// the pair of cursor indices sorted least to greatest. When transforming a
+// cursor with no selection, mark should equal index.
+type TextCursor struct {
+	index int
+	mark  int
+}
+
+// NewTextCursor returns a new TextCursor spanning the given indices. If mark equals
+// index, the cursor will have no selection.
+func NewTextCursor(index int, mark int) TextCursor {
+	return TextCursor{index, mark}
+}
+
+// Index returns the cursor index.
+func (c TextCursor) Index() int {
+	return c.index
+}
+
+// Mark returns the cursor mark index.
+func (c TextCursor) Mark() int {
+	return c.mark
+}
+
+// Range returns the difference between indices ordered highest and lowest.
+func (c TextCursor) Range() int {
+	if c.index > c.mark {
+		return c.index - c.mark
+	}
+	return c.mark - c.index
+}
+
+// Selection returns the cursor indices ordered least to greatest.
+func (c TextCursor) Selection() (int, int) {
+	if c.index < c.mark {
+		return c.index, c.mark
+	}
+	return c.mark, c.index
+}
+
+// HasSelection returns whether this cursor spans a range greater than zero.
+func (c TextCursor) HasSelection() bool {
+	return c.Range() > 0
+}
+
+// Invert returns the cursor in opposite order.
+func (c TextCursor) Invert() TextCursor {
+	return TextCursor{c.mark, c.index}
+}
+
+// Offset returns the cursor with indices + i.
+func (c TextCursor) Offset(i int) TextCursor {
+	return TextCursor{c.index + i, c.mark + i}
+}
+
+// NewSelection returns a cursor oriented as c with given selection.
```

Comment:
providing a bool to orient a `TextCursor` worries me. This ended up with some strange abuses in the old code that made things hard to follow and grasp, particularly with the lack of rigor in how start and end were set in combination with the use of a bool during transforms.

At the same time, I don't necessarily like having this as a method on TextCursor but stopped short of making a new function as I couldn't think of a name I liked and feeling like a low priority at the time.

I've noted the idea of an orientation in some of the docs when you have the logical components of a cursor and a selection, but calling `cursor.NewSelection` as you say isn't immediately obvious why you would do that since the cursor itself is only used to inpect an orientation for the creation of something new not related to it.

The idea of an orientation could become a concrete type like

```go
type Orientation int

const (
  Left Orientation = iota
  Right
)
```

with a method to retrieve orientation and making the distinction more clear, but I didn't see the value in that at the time. But perhaps it does make sense? Function names get a bit silly at this point though, for example `NewTextCursorSelectionOrient(start int, end int, orient Orientation) TextCursor`